aws/src/main/java/org/apache/iceberg/aws/AwsClientCredentialsFactory.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.aws;
+
+import java.io.Serializable;
+import org.apache.iceberg.catalog.CatalogConfigurable;
+import software.amazon.awssdk.awscore.client.builder.AwsClientBuilder;
+
+/**
+ * Interface for loading a custom factory to configure AWS credentials for different services.
+ * See {@link AssumeRoleCredentialsFactory} as an example.
```

Comment:
Can you give a quick example of how this is used and how it simplifies configuration?

Reply:
The `AssumeRoleCredentialsFactory` is an example for it, I can give another example with more complicated customer use case:

```java
package com.my.team;

import software.amazon.awssdk.auth.credentials.ContainerCredentialsProvider;
import software.amazon.awssdk.awscore.client.builder.AwsClientBuilder;
import software.amazon.awssdk.services.glue.GlueClient;
import software.amazon.awssdk.services.glue.GlueClientBuilder;
import software.amazon.awssdk.services.s3.S3ClientBuilder;
import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
import software.amazon.awssdk.services.sts.auth.StsAssumeRoleWithSamlCredentialsProvider;

public class CustomCredentialsFactory implements AwsClientCredentialsFactory {


  @Override
  public void configure(AwsClientBuilder clientBuilder) {
   if (clientBuilder instanceof S3ClientBuilder) {
     clientBuilder.credentialsProvider(StsAssumeRoleWithSamlCredentialsProvider.builder()
         .refreshRequest(...)
         .build());
   } else if (clientBuilder instanceof GlueClientBuilder) {
     clientBuilder.credentialsProvider(StsAssumeRoleCredentialsProvider.builder()
         .refreshRequest(...)
         .build())
   } else {
     clientBuilder.credentialsProvider(ContainerCredentialsProvider.builder().build());
   }
  }
}
```

And it can be started with:
```
spark-sql --jars iceberg-spark3-runtime.jar \
    --conf spark.sql.catalog.my_catalog=org.apache.iceberg.spark.SparkCatalog \
    --conf spark.sql.catalog.my_catalog.catalog-impl=org.apache.iceberg.aws.glue.GlueCatalog \
    --conf spark.sql.catalog.my_catalog.warehouse=s3://my-bucket \
    --conf spark.sql.catalog.my_catalog.client.credentials-factory=com.my.team.CustomCredentialsFactory
```

Reply:
So, I'm just catching up on all of the discussion and may not be following everything just yet, but I feel like we should try to avoid most S3Client configuration in Iceberg.  Part of the reason the S3FileIO takes a supplier for an S3Client is the we can externalize all of the configuration.

The places where we have exposed configuration (part size, acl, etc.) is because they are request level options.  Since credentials are at the client level, we should be able to have just a S3ClientFactory that supplies all of the custom client configuration as opposed to exposing it directly. 

I feel like that would really reduce the complexity in Iceberg and at the same time expose all of the possible configuration for the S3Client that anyone may want to leverage.

Reply:
This PR mostly comes from the following use cases:
1.  use a customized credential provider across all the AWS clients
2. use some kind of proxy configuration for HTTP client to bypass network boundary (not included in this PR, plan to put in another one)

I completely agree that we should expose limited number of configurations for S3FileIO, but those are not related to S3 but AWS clients in general.

And maybe I missed something, but based on my understanding, for people using open source Iceberg directly, it is not straightforward to use the supplier, unless they do something like
```java
public MyS3ClientFileIO extends S3FileIO {
  public MyS3ClientFileIO() {
    super(new MyS3ClientSupplier());
  }
}
```

And load the `MyS3ClientFileIO` instead of `S3FileIO` as the FileIO implementation, which is (1) not very convenient, and (2) people with the requirements I listed basically have to override all the classes that touches AWS clients.

Reply:
@jackye1995  I'm not proposing that users would subclass the S3FileIO, but rather that we take your approach of dynamically loading and just up-level it to the whole S3Client supplier as opposed to just parts of the client configuration.  This would mean configuring with something like `--conf spark.sql.catalog.my_catalog.client.s3.supplier=com.my.team.S3ClientSupplier`.  We could then just change the S3FileIO to check for a configured supplier before using the default client.

This allows anyone to supply a fully configured client based on their own requirements and we wouldn't need to expose the specific configurations through the Iceberg APIs.

Reply:
I see, yes I have been personally debating about this and that's why I put up this PR for feedbacks. We can either as you suggest provide integration points to load each AWS client, or in my current PR to provide integration points to load important aspects like credential providers, and there are pros and cons for both.

Currently what I see is that we will at least have the following clients in the AWS module:
- Glue
- DynamoDB
- S3
- KMS
- (planned) CloudWatch
- and might be more

And in my use cases listed, it is much cleaner to just expose a holistic factory for credentials rather than to load every single AWS client dynamically.

So what about exposing a single factory named `AwsClientConfigurationFactory`, for people to configure any client configuration? It will look like:

```java
public interface AwsClientConfigurationFactory implements CatalogConfigurable, Serializable {

    <T extends AwsClientBuilder & AwsSyncClientBuilder> T configure(T clientBuilder);
}

public class MyClientConfigurationFactory implements AwsClientConfigurationFactory {

    <T extends AwsClientBuilder & AwsSyncClientBuilder> T configure(T clientBuilder) {
        if (clientBuilder instanceof S3ClientBuilder) {
             S3ClientBuilder s3Clientbuilder = (S3ClientBuilder) clientBuilder;
             // do whatever you want with s3
       }
      // do common stuffs across all client builders
       clientBuilder.credentialsProvider(...);
    }
}
```

By doing so, there is only 1 integration point for all AWS clients, and it can satisfy all both use case for configuring a single client, or configuring common properties across all clients.

Reply:
@danielcweeks see updated PR for more details.